File: hackerrank/30 days of code/day 12, inheritance/inheritance.cs
// https://www.hackerrank.com/challenges/30-inheritance/problem?isFullScreen=true

using static System.Console;


public static class Solution
{
    public static void Main()
    {
        var (personalData, studentData) = ReadInput();
        var student = new Student(personalData, studentData);
        student.PrintPerson();
        WriteLine("Grade: " + student.CalculateGrade());
    }

    private static (PersonalData, StudentData) ReadInput()
    {
        var line = ReadLine().Split();
        (string firstName, string lastName, int id) = (line[0], line[1], int.Parse(line[2]));
        ReadLine(); // Skip the next line
        int[] scores = [.. ReadLine().Split().Select(int.Parse)];
        return (new PersonalData(firstName, lastName, id), new StudentData(scores));
    }
}

public record PersonalData(string FirstName, string LastName, int Id);
public record StudentData(int[] Scores);

public class Person(PersonalData data)
{
    private string _firstName = data.FirstName, _lastName = data.LastName;
    private int _id = data.Id;

    public void PrintPerson()
    {
        WriteLine("Name: " + _lastName + ", " + _firstName + "\nID: " + _id);
    }
}

public class Student(PersonalData personalData, StudentData studentData) : Person(personalData)
{
    private int[] _scores = studentData.Scores;

    public char CalculateGrade()
    {
        int avg = _scores.Sum() / _scores.Length;

        if (avg > 100 || avg < 40)
            return 'T';
        else if (avg >= 90)
            return 'O';
        else if (avg >= 80)
            return 'E';
        else if (avg >= 70)
            return 'A';
        else if (avg >= 55)
            return 'P';
        return 'D';
    }
}

File: hackerrank/30 days of code/day 12, inheritance/inheritance.ts
// https://www.hackerrank.com/challenges/30-inheritance/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString: string = '';
let inputLines: string[] = [];
let currentLine: number = 0;

process.stdin.on('data', function (inputStdin: string): void {
  inputString += inputStdin;
});

process.stdin.on('end', function (): void {
  inputLines = inputString.split('\n');
  inputString = '';
  main();
});

function readLine(): string {
  return inputLines[currentLine++];
}

//////////////////////////////////////////////////

function main() {
  let [personalData, studentData] = readInput();
  let student = new Student(personalData, studentData);
  student.printPerson();
  console.log('Grade: ' + student.calculateGrade());
}

function readInput(): [PersonalData, StudentData] {
  let [firstName, lastName, id] = readLine().split(' ');
  readLine(); // Skip next line about size of scores
  let scores: number[] = readLine().split(' ').map(Number);
  return [{ firstName, lastName, id }, { scores }];
}

interface PersonalData {
  readonly firstName: string;
  readonly lastName: string;
  readonly id: string;
}

interface StudentData {
  readonly scores: number[];
}

class Person {
  private firstName: string;
  private lastName: string;
  private id: string;

  constructor(personalData: PersonalData) {
    this.firstName = personalData.firstName;
    this.lastName = personalData.lastName;
    this.id = personalData.id;
  }

  printPerson() {
    console.log(
      'Name: ' + this.lastName + ', ' + this.firstName + '\nID: ' + this.id,
    );
  }
}

class Student extends Person {
  private scores: number[];

  constructor(personalData: PersonalData, studentData: StudentData) {
    super(personalData);
    this.scores = studentData.scores;
  }

  public calculateGrade(): string {
    let sum: number = this.scores.reduce((a, b) => a + b, 0);
    let avg: number = sum / this.scores.length;

    if (avg > 100 || avg < 40) return 'T';
    else if (avg >= 90) return 'O';
    else if (avg >= 80) return 'E';
    else if (avg >= 70) return 'A';
    else if (avg >= 55) return 'P';
    return 'D';
  }
}
